index.src.html

```diff
@@ -2394,14 +2420,812 @@
+    means running these steps:
+    1. Let |supported| be the result of running the [=ImageDecoder/Check Type
+        Support=] algorithm with `init.type`.
+    2. If |supported| is `false`, queue a task on the [=control thread=] event
+        loop to run the [=ImageDecoder/Close ImageDecoder=] algorithm
+        with a {{NotSupportedError}} {{DOMException}} and abort
+        these steps.
+    3. If |supported| is `true`, assign the
+        {{ImageDecoder/[[codec implementation]]}} internal slot with an
+        implementation supporting `init.type`
+    4. Configure {{ImageDecoder/[[codec implementation]]}} in accordance with
+        the values given for {{ImageDecoderInit/premultiplyAlpha}},
+        {{ImageDecoderInit/colorSpaceConversion}},
+        {{ImageDecoderInit/desiredWidth}}, and
+        {{ImageDecoderInit/desiredHeight}}.
+
+### Attributes ### {#imagedecoder-attributes}
+
+: <dfn attribute for=ImageDecoder>tracks</dfn>
+:: Provides metadata for the available tracks and a mechanism for selecting a
+    track to decode.
+
+    The {{ImageDecoder/tracks}} getter steps are to return
+    {{ImageDecoder/[[ImageTrackList]]}}.
+
+: <dfn attribute for=ImageDecoder>complete</dfn>
+:: Indicates whether {{ImageDecoder/[[encoded data]]}} is completely buffered.
+
+    The {{ImageDecoder/complete}} getter steps are to return
+    {{ImageDecoder/[[complete]]}}.
+
+
+### Methods ### {#imagedecoder-methods}
+: <dfn method for=ImageDecoder>decodeMetadata()</dfn>
+:: Returns a promise which resolves once enough
+    {{ImageDecoder/[[encoded data]]}} has been buffered and decoded to
+    establish {{ImageDecoder/tracks}}.
+
+    NOTE: {{ImageTrack}} {{ImageTrack/frameCount}} may receive subsequent
+        updates until {{ImageDecoder/complete}} is `true`.
+
+    When invoked, run these steps:
+    1. If {{ImageDecoder/[[closed]]}} is `true`, return a {{Promise}}
+        rejected with an {{InvalidStateError}} {{DOMException}}.
+    2. If {{ImageDecoder/[[tracks established]]}} is `true`, return a resolved
+        {{Promise}}.
+    3. Let |promise| be a new {{Promise}}.
+    4. Append |promise| to {{ImageDecoder/[[pending metadata promises]]}}
+    5. If {{ImageDecoder/[[decode metadata requested]]}} is `false`:
+        1. [=Queue a control message=] to [=ImageDecoder/decode track
+            metadata=].
+        2. Assign `true` to {{ImageDecoder/[[decode metadata requested]]}}.
+    5. Return |promise|.
+
+    [=Running a control message=] to <dfn for=ImageDecoder>decode track
+    metadata</dfn> means running these steps:
+    1. Run the [=ImageDecoder/Establish Tracks=] algorithm.
+
+: <dfn method for=ImageDecoder>decode(options)</dfn>
+:: Enqueues a control message to decode the frame according to |options|.
+
+    When invoked, run these steps:
+    1. If {{ImageDecoder/[[closed]]}} is `true`, return a {{Promise}}
+        rejected with an {{InvalidStateError}} {{DOMException}}.
```

Comment:
TODO: consider throwing a NotSupportedError here if the reason for close was unsupported mime type. Generally, I think we should save the exception that was used when running the close algorithm and re-use in cases such as this.

Reply:
I've had second thoughts. Purely a style thing, but in hindsight I like that the pending promises are all rejected with whatever the initial reason is with follow up operations being rejected because the codec is now in an invalid state (i.e. "closed")